scalacheck/src/main/scala/pathy/scalacheck/GenPath.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2014 - 2015 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pathy
+package scalacheck
+
+import Path._
+
+import org.scalacheck.Arbitrary
+import org.scalacheck.Gen
+
+import scalaz.Show
+import scalaz.syntax.show._
+
+/** Represents a `Path[B,T,S]` indexed by another type `A` which is used to
+  * generate path segments.
+  *
+  * i.e. For some type `A` such that Arbitrary[A] and Show[A],
+  * Arbitrary[GenPath[B,T,S,A]] will generate an arbitrary Path[B,T,S] where
+  * every path segment is formed by Arbitrary[A].shows.
+  */
+final case class GenPath[B,T,S,A](path: Path[B,T,S]) extends AnyVal
```

Comment:
Playing devil's advocate, a `ConstrainedPath` type might encourage someone to use it in a context other than arbitrary value generation and might be led to believe that a `ConstrainedPath` might offer some kind of guarantee that a path is free of certain Characters when that is not the case, it's just a marker trait for `Path` generation. So in light of that, it might make sense to include that intent in the data name.

However, maybe this is actually a generally useful thing. Maybe it might make sense to support restricting the composition of `Path`s. I can certainly conceive of a situation where a system will only accept paths that are free of certain characters and it would be useful to reason about such things.